bindings/java/devdoc/java_out_of_process_hld.md

```diff
@@ -0,0 +1,122 @@
+Java Out of Process Module API
+==============================
+
+Overview
+--------
+
+This document describes the API that enables communication between an out of process Java module and Azure IoT Gateway. 
+An out of process module should implement `IGatewayModule` interface or extend the `GatewayModule` abstract class as it does for an in-process module, more details in [Java binding high level design](java_binding_hld.md). In addition, the API provides `GatewayModuleProxy` which is a proxy for the IGatewayModule implementation and handles the communication between the Gateway and the out of process module. 
+`GatewayModuleProxy` contains the methods to *attach* a module to the Gateway, *listen* for messages and *detach* from the Gateway. 
+
+There are two types of messages that are exchanged between the Gateway and the out of process module:
+
+-  *control* messages that are used to signal `CREATE`, `START` and `DESTROY` of a module. These messages are always send from the Gateway to the module.
+-  *data* messages which are send between modules via the Gateway.
+
+
+![](java_oop_module_hld.png)
+
+Receive out of process messages from the Gateway
+------------------------------------------------
+
+In order to make a module to run out of process, create an instance of `GatewayModuleProxy`, pass the configuration that contains the unique identifier to attach to the Gateway and then call `attach` and `startListening`. 
+Also the Gateway configuration has to be changed for that specific module to set the loader as an out of process loader and set the same unique identifier. 
+
+``` java
+interface ModuleProxy {
+	public void attach();
+	public void detach();
+	public void startListening() ;
+	public void stopListening();
+	public void receiveMessage();
+}
+```
+
+``` java
+public class GatewayModuleProxy implements ModuleProxy {
+	private ModuleConfiguration config;
+	public GatewayModuleProxy(ModuleConfiguration config) {
+		// implementation
+	}
+
+	@Override
+	public void attach() {
+		// implementation	
+	}
+	
+	@Override
+	public void detach() {
+		// implementation	
+	}
+
+	@Override
+	public void receiveMessage() {
```

Comment:
Then I would have only `OopModule_Attach` and `OopModule_Detach`. This would allow us to have the "same" API, only the Java would be smaller as a function of the language.